TYP: Make array _ShapeType bound and covariant

Jacob-Stevens-Haas opened the pull request:
Fixes #25729.  Applies to main ndarray class as well as record arrays, masked arrays, masked record arrays, mmaps, and matrices.

I didn't see anything in the dev guide about testing static types for PRs, but this change fixes the MWE in the linked issue.  While there hasn't been much discussion on that issue, #16544 has renewed interest, so submitting this in case it's  helpful.

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

charris commented:
@BvB93 Ping.

jorenham commented on doc/release/upcoming_changes/README.rst:
```diff
@@ -40,7 +40,7 @@ So for example: ``123.new_feature.rst`` would have the content::
     To use it, write ``np.my_favorite_function(..., my_new_feature=True)``.
 
-``highlight`` is usually formatted as bulled points making the fragment
+``highlight`` is usually formatted as bullet points making the fragment
```
🐮

BvB93 commented on doc/release/upcoming_changes/26081.improvement.rst:
```diff
@@ -0,0 +1,21 @@
+Shapes are now covariant
+------------------------
+When ``ndarray`` was made generic, more details and implementation were
+left to future development.  While it was recognized that shape was inherently
+a tuple (and thus covariant), development towards ``TypeVarTuple`` does not
+yet allow bounds or variance.  This change allows user code to type check in
+functions that attempt to specify the shape parameter:
```
This comment about the type bounds is not quite accurate in the case of `ndarray`. The ndarray class was made generic prior to the development of `TypeVarTuple`, and as such what was effectively a placeholder slot was added for a future shape `TypeVar`. And while type var tuples do not support a bound, type vars *do*:

```diff
- _ShapeType = TypeVar("_ShapeType", covariant=True, bound=Any)
+ _ShapeType = TypeVar("_ShapeType", covariant=True, bound=tuple[int, ...])
```

Another option is to convert the first type var of `ndarray` into a type var tuple, but this has the very drawback that (to the extent of my knowledge) there's no way of representing a tuple with an unknown/arbitrary length (aka array dimensions).

Jacob-Stevens-Haas replied:
Thank you for the review!

> The `ndarray` class was made generic prior to the development of `TypeVarTuple`

Oh, sorry, I was just going off of git blame vs PEP date, and trying to fill in some context for readers who had heard of `TypeVarTuple` why this PR was moving away from that option:  changing `_ShapeType` to covariant pre-empts a PR changing it to a `TypeVarTuple` until the latter supports variance.  There's also more of my rationale at the end of this comment.  Would the following language work?

```
Static shape typing for `ndarray` is a long-term effort that continues 
with this change.  While shape is a variadic type variable, this change
chooses to keep `TypeVar` rather than `TypeVarTuple`, as the latter
does not yet allow bounds or variance.  This choice allows users
to type check functions that attempt to specify the shape parameter
as a `NewType`:
```

> And while type var tuples do not support a bound, type vars _do_

I'm 99% on board with bounding to `tuple[int, ...]`, as in your code and @jorenham's suggestion.  My only reservation is that I don't know enough about structured array data types.  E.g., in a length-3 structured array with two fields (or a record array of the same), would the shape typevar be `tuple[int, int]`?  Or is the dtype typevar responsible for the field length, meaning the shape typevar is `tuple[int]`?  Because I don't know, I limited this PR just towards variance, not bounds.  But if it's desired, I'm more than happy to add the bound.

> Another option is to convert the first type var of ``ndarray`` into a type var tuple, but this has the very drawback that (to the extent of my knowledge) there's no way of representing a tuple with an unknown/arbitrary length (aka array dimensions).

I thought `TypeVarTuple`s are always arbitrary length before they're narrowed/resolved?  But by gut feeling alone, `TypeVarTuple` seem less useful than `TypeVar("Shape", bound=tuple[int, ...])` until more of the TBDs from PEP 646 are resolved (e.g. [manipulating individual types][1] and [shape arithmetic][2]), especially because `Tuple` now also supports unpacking types.  I also prefer the behavior of `tuple` to `TypeVarTuple` when dealing with [unknown type parameters][3].  That is, I prefer type checking to show more shape errors, even if it requires peppering my code with `cast`.

[1]: https://peps.python.org/pep-0646/#overloads-for-accessing-individual-types
[2]: https://peps.python.org/pep-0646/#shape-arithmetic
[3]: https://peps.python.org/pep-0646/#unspecified-type-parameters-tuple-vs-typevartuple

BvB93 commented on doc/release/upcoming_changes/26081.improvement.rst:
```diff
@@ -0,0 +1,21 @@
+   Time = NewType("Time", int)
+   Series = NewType("Series", int)
+
+   arr: np.ndarray[tuple[Time, Series], Any] = np.arange(4.0).reshape((2, 2))
```
I'd strongly prefer to add another type alias for shaped arrays to the `numpy.typing` namespace instead. Using plain `np.ndarray` as a type annotation (over `npt.NDArray`) is generally ill-advised due to the former's complexity, if it can be avoided that is. Something like a `ShapedArray` would be ideal. _i.e._ a type alias representing an array with a given shape *and* dtype; the docs should explicitly recommend `npt.NDArray` usage if the shape is unknown/arbitrary.

Note that prior to Python 3.11 people would have to make sure during runtime to either use `if TYPE_CHECKING:` blocks and/or `from __future__ import annotations` due to a lack of `TypeVarTuple` support. This caveat is perfectly acceptable though in my opinion.

```python
_ShapeType = TypeVarTuple("_ShapeType")
_ScalarType_co = TypeVar("_ScalarType_co", bound=generic, covariant=True)

if TYPE_CHECKING or sys.version_info >= (3, 11):
    ShapedArray = ndarray[tuple[*_ShapeType], dtype[_ScalarType_co]]
else:
    ShapedArray = ndarray[Any, dtype[_ScalarType_co]]
```

jorenham replied:
@BvB93 I proposed exactly solution a couple of days ago https://github.com/numpy/numpy/issues/16544#issuecomment-2005169857. Great minds really do think alike, eh 😉 ?

Jacob-Stevens-Haas replied:
Happy to add an alias!  My only consideration is whether the alias should expect grouping dimensions, e.g.:
```python
FloatArray2D = npt.ShapedArray[tuple[int, int], np.float64]
```
I slightly agree with @vnmabus's [comment](1): "I think it [requiring dimensions to be grouped] would make it more consistent with functions and classes that need to take more than one shape".  That said, I'll defer to you if you want the flatter version of `ShapedArray`?

[1]: https://github.com/numpy/numpy/issues/16544#issuecomment-2006220209

jorenham replied:
I don't think it should.

In my experience, it is very annoying to have to deal with long typing annotations. In this case, it is a differnce of 7 characters; close to 10% of the commonly 80 or 79 character width limit. To illustrate:

```python
npt.ShapedArray[tuple[int, int], np.float64]
npt.ShapedArray[int, int, np.float64]
```

If you still want to use "group" the shape types, then there's always  `np.ndarray[tuple[int, int], np.dtype[np.float64]]` to fall back to.

jorenham replied:
But for the sake of readability, it might help to place the shapes *after* the dtype, e.g.

```python
type VectorF8[I: int] = npt.ShapedArray[np.float64, I]
type MatrixF8[I: int, J: int] = npt.ShapedArray[np.float64, I, J]
type CuboidF8[I: int, J: int, K: int] = npt.ShapedArray[np.float64, I, J, K]
```

I realize that this doesn't match the order of the `dtype` and `shape` arguments in e.g. `np.array`, but it might be more intuitive this way, since those used to `npt.NDArray` have it in their muscle memory to always start with the dtype. 

Furthermore, Python users are used to always having the `*args` after any positional arguments. So doing something different with the `*Shape` type parameter could be confusing.

jorenham replied:
On a related note; what about naming it `Array` instead of `ShapedArray`?

Jacob-Stevens-Haas replied:
I think there's going to be a lot of opinions on variadic shape vs tuple shape and shape first vs dtype first.  Here's the pros/cons I've thought through so far:

#### variadic shape, dtype first:
* Follows example in PEP 646
* Similar to `foo(arg1, *args)`
  * People using `NDArray` are used to dtype first
* Shorter
#### tuple shape, dtype last:
* Similar to funcs like `np.ones(shape: tuple[int, ...], dtype: dtype)`
  * People who haven't used static typing are used to this order
* Same as `ndarray` typing
  * People who've already been static shape typing have to use this syntax
 
Polled my office and people unanimously supported the second.

**variadic shape, dtype last** seems like a decent compromise, and it's what I'll commit. I'm happy to sit on the PR and change it back if a better consensus is desired, but I've overthought myself to a point of analysis paralysis.  I've coded all these options up, hopped back and forth, and test-drove them in a few use cases (and they all seem to work fine)

> On a related note; how about naming it Array instead of ShapedArray?


Haha yeah good point.  ~thefacebook.com~ facebook.com :smile:

Jacob-Stevens-Haas replied:
Also a neat lesson learned here: 
```python
TupType = TypeVar("TupType", covariant=True, bound=tuple[int, ...])
TypeTup = TypeVarTuple("TypeTup")

class Foo[TupType]: ...
Bar = Foo[tuple[*TypeTup]]

def myfunc(a: Bar): ...
```
^ an effective way of enforcing type variance and bounds on a `TypeVarTuple`.  Combines (some of) the advantages of both.

Jacob-Stevens-Haas replied:
Actually, the unpacking restrictions on `tuple[int, ...]` mean that the third option isn't usable for certain functions, e.g. `def add_one_dimension(arr: Array[*Ts, Any]) -> Array[int, *Ts, Any]`

jorenham replied:
It might help to return e.g. `Array[Any, *Ts, np.float64]`, instead of `Array[int, *Ts, np.float64]`. I'm not sure why, but `Any` results in covariant behaviour, whereas `int`  doesn't.

I agree that it would be better if you could `TypeVarTuple` as `covariant=True`, and `bound=int`, but unfortunately that's not implemented at the moment (for no valid reason whatsoever IMHO).

BvB93 replied:
> variadic shape, dtype last seems like a decent compromise, and it's what I'll commit. I'm happy to sit on the PR and change it back if a better consensus is desired, but I've overthought myself to a point of analysis paralysis. I've coded all these options up, hopped back and forth, and test-drove them in a few use cases (and they all seem to work fine)

I'm personally in favor of this approach, yes. The main goal of this type alias is to provide a _convenient_ alias for the end-user that's less verbose than the full syntax, providing syntactic sugar for turning something like `np.ndarray[tuple[int, int], np.dtype[np.float64]` into a more manageable `npt.Array[int, int, np.float64]`. 

Are there edge cases wherein this alias cannot be used? As you have noted previously: yes, yes there are. Is this a problem? No, not as long as it's usable in the majority of cases; in all others one can always fall back to plain` np.ndarray` usage (something which holds as well for `npt.NDArray`).

BvB93 reviewed (dismissed):
This is a good change I would say, though one thing I feel is still really missing is a convenient public-facing type alias for shaped arrays (see comment below).

Jacob-Stevens-Haas commented:
Alright @BvB93 , I made all the requested changes and CI is green - ready for another review. (a) Added type alias, (b) bound the TypeVar, (c) to demonstrate the use, `np.shape` is now typed to return the shape `TypeVar`, (d) described this in the docs and in expanded change notes, including some of the interplay between `TypeVar` and `TypeVarTuple`.  Hope I've been clear enough without going into too much detail.

Big thanks to @jorenham for being equally interested in this and helping me think through some of the details!

mtsokol reviewed (commented):
I think we could have a few reveal tests for the new `Array` type, in one of the `numpy/typing/tests/data/reveal/` files (or a new test file for that), something similar to:

```python
if sys.version_info >= (3, 11):
    Length = NewType("Length", int)
    Width = NewType("Width", int)
    arr: npt.Array[Length, Width, np.int8]
    assert_type(arr, npt.Array[int, int, np.int8])
```

jorenham commented on doc/release/upcoming_changes/26081.improvement.rst:
```diff
@@ -0,0 +1,44 @@
+``ndarray`` Shapes are now covariant and bound
+---------------------------------------------------------------
+Static shape typing for ``ndarray`` is a long-term effort that continues 
+with this change.  Array shape can be described with a tuple of ints, but
```
Isn't this the first explicit implementation of shape typing?

jorenham commented on doc/release/upcoming_changes/26081.improvement.rst:
```diff
@@ -0,0 +1,44 @@
+---------------------------------------------------------------
+Static shape typing for ``ndarray`` is a long-term effort that continues 
+with this change.  Array shape can be described with a tuple of ints, but
+both ``TypeVar`` and ``TypeVarTuple`` provide some useful functionality.
+The former allows bounds and type variance, whereas the latter allows
+unpacking.  ``ndarray`` is built with a ``TypeVar``, and this change
+provides the appropriate bounds and variance.  It allows users
```
`TypeVar` and `TypeVarTuple` are implementation details, as they won't be needed anymore once support for python versions below 3.12 has been dropped.
I believe that the implementation-agnistic terms *type parameter* and *variadic type parameters* are better suited here.

Jacob-Stevens-Haas replied:
Good point.  I was trying to include too much of the PR/development thought process in the release notes, but rereading this a week later it seems forced.  I'll take another go at it.

Jacob-Stevens-Haas commented:
> I think we could have a few reveal tests for the new Array type

Ok, so I added some tests for `ndarray`, and `npt.Array`, but I put them in `test_success` rather than `test_reveal` - I was unsure of the difference.  Larger problem, the `Array` ones fail.  The problem is that `Array` unpacks a `TypeVarTuple` into a bound `TypeVar`.  pyright doesn't see a problem with it, but mypy does.  The non-tuple equivalent:

```python
from typing import TypeVar

T = TypeVar("T", bound=str)
S = TypeVar("S")

def foo(a: T) -> T: ...
def bar(b: S, c: S) -> T:
    return foo(b)  # type error since object is not a str
```

Unfortunately, pyright also accepts things like `npt.Array[str, str, np.uint8]`

As a result, I've reverted the tests for `npt.Array` (so that people can play with it if they'd like), but left in `npt.Array` as written.  It seems like, **until `TypeVarTuple` supports binding and variance, there's two options**:
a) Change `npt.Array` to take a [tuple shape, not a variadic one](https://github.com/numpy/numpy/pull/26081#discussion_r1535554708).
b) Decide the current API is desired, ship as currently written, refer bug reports to `np.ndarray`, and use the experience to motivate further work on PEP 646.

jorenham commented:
@Jacob-Stevens-Haas I agree that it would be best if there'd be upper bounds on the variadic shape type params. 
But I don't see why not having them is a hard requirement. Because if you do so anyway, it should be possible for type checkers to detect that it's invalid nonetheless. And just because there are typecheckers that currently don't implement this, doesn't mean that they never will. 
I even bet that once people start using `numpy.typing.Array[*Shape, DType]`, type checkers will quickly implement such a feature. So it's a chicken & egg story, but someone has to lay the egg first 😉 .

Jacob-Stevens-Haas commented:
@jorenham if you're talking about option (b), which means I don't have to do any extra work, I'm happy to lay that :egg: haha.  But if it's not a type checker issue and actually incorrect according to type theory (under the limits of what PEP 646 provides), I kinda want an established numpy maintainer to weigh in.

jorenham commented:
@Jacob-Stevens-Haas Yes, option B seems to be the most obvious to me. 
In several of my projects I've been using effectively the same variadic alias, with sub-aliases for specific common `ndim`, e.g. 

```
type Vector[N: int, T: np.generic] = Array[N, T]
type Matrix[M: int, N: int, T: np.generic] = Array[M, N, T]
```

I'm not saying you should add these here, but it's just an example of how you can add upper bound restrictions on shape type parameters, even if they are missing from `Array`.

Jacob-Stevens-Haas commented:
@BvB93 @mtsokol, I've implemented your requested changes, are you able to re-review?

Jacob-Stevens-Haas commented:
Trying to tempt fate by publicly posting "I can't figure this out" while I try this one more time:

So I'm coming to believe that any form of `npt.Array` that uses `TypeVarTuple` may have to be delayed for another PR.  I'm chasing around failed tests, and while things are hunky dory for 3.11 and 3.12, something always goes wrong with 3.10

* Typing extensions is unavailable in some CI environments, meaning `npt.Array` has to be assigned a different value in <3.11
* Typing tests are executed in 3.10, 3.11, and 3.12.  Even if we guard shape311.py with `if sys.version_info ...`, 3.10 still checks those files
* If it can't handle a variadic number of parameters, for Array, it fails when it tries to run the tests (separate from testing type checking, all of the test files need to run)
* For some reason, even with quoted types (delayed evaluation of annotations) it still fails when `Array = NDArray`

BvB93 commented on numpy/_typing/_add_docstring.py:
```diff
@@ -149,4 +149,45 @@ def _parse_docstrings() -> str:
+    alias :term:`generic <generic type>` w.r.t. its shape and
+    `dtype.type <numpy.dtype.type>`.
+
+    Can be used during runtime for typing arrays with a given dtype
```
I'd drop the "during runtime" part, as the statement is equally valid both in .py and .pyi files

BvB93 commented on numpy/_typing/_add_docstring.py:
```diff
@@ -149,4 +149,45 @@ def _parse_docstrings() -> str:
+    Can be used during runtime for typing arrays with a given dtype
+    and specified shape.  Particularly useful for functions which modify number
+    of axes.  If it is important to enforce integer axis sizes, use
+    `np.ndarray <numpy.ndarray>` for typing.  If axes dimensions are
```
```suggestion
    `np.ndarray <numpy.ndarray>` for typing.  If axes dimensions are
```
I'd maybe drop the reference to `np.ndarray` and the mention of enforcing integer axis sizes. This feels a bit too much like discussing internal implementation details for it to belong in the general docs (well, those associated with upstream limitations anyway...).

BvB93 commented on numpy/_typing/_add_docstring.py:
```diff
@@ -149,4 +149,45 @@ def _parse_docstrings() -> str:
+    and specified shape.  Particularly useful for functions which modify number
+    of axes.  If it is important to enforce integer axis sizes, use
+    `np.ndarray <numpy.ndarray>` for typing.  If axes dimensions are
+    variable, it is best to use `npt.NDArray <numpy.typing.NDArray>`.
```
```suggestion
    unknown and/or irrelevant, it is best to use `npt.NDArray <numpy.typing.NDArray>`.
```
This doesn't sound quite right: axis variability is by no means guaranteed to be a problem here, as typevars/typevartuples can keep track of those just fine (in theory). It's really when the array shape is either unknown or (completely) irrelevant that there's no reason _not_ to use `NDArray`.

Jacob-Stevens-Haas replied:
Doing this, but is it worth leaving any hint that practice hasn't necessarily caught up to theory yet? (mypy still doesn't like `Array[*Shape, dtype]`, but it looks like pyright now accepts this)

BvB93 commented on numpy/_typing/_add_docstring.py:
```diff
@@ -149,4 +149,45 @@ def _parse_docstrings() -> str:
+        >>> Length = NewType("Length", int)
+        >>> Width = NewType("Width", int)
+        >>> GreyscaleImage = npt.Array[Length, Width, np.int8]
+        >>> a: GreyscaleImage = np.arange(100, dtype=np.int8).reshape((10,10))
```
```suggestion
        >>> a: GreyscaleImage = np.arange(100, dtype=np.int8).reshape((10, 10))
```
Minor pep8 nit.

BvB93 commented on numpy/_typing/_add_docstring.py:
```diff
@@ -120,7 +120,7 @@ def _parse_docstrings() -> str:
+    A `np.ndarray[Any, np.dtype[+ScalarType]] <numpy.ndarray>` type alias
     :term:`generic <generic type>` w.r.t. its `dtype.type <numpy.dtype.type>`.
 
     Can be used during runtime for typing arrays with a given dtype
```
I'd append a sentence here with something along the line of "See `npt.NDArray <numpy.typing.NDArray>` for dealing with arrays of a specified shape."; something to easily allow one to find the `npt.Array` docs starting from `npt.NDArray`.

BvB93 commented on numpy/_typing/_add_docstring.py:
The `ndarray.__class_getitem__` [docs](https://github.com/numpy/numpy/blob/ac5614f72afef317f6c0eec66d0d7f90900bf863/numpy/_core/_add_newdocs.py#L2994-L3021) could also do with a new reference to `npt.Array`; something to make the docs as easily findable as possible:
```
    See Also
    --------
    ...
    numpy.typing.Array : An ndarray alias :term:`generic <generic type>` w.r.t. 
        both its `shape <numpy.ndarray.shape>` and `dtype.type <numpy.dtype.type>`.
```

BvB93 commented on numpy/typing/tests/data/fail/shape311.py:
```diff
@@ -0,0 +1,9 @@
+from typing_extensions import assert_type
+
+# Fails for different reason with same syntax < 3.11
+npt.Array[str, str, np.float64]   # E: Value of type variable
```
```suggestion
npt.Array[str, str, np.float64]   # E: Value of type variable

```
Linting nit

BvB93 commented on doc/release/upcoming_changes/26081.improvement.rst:
```diff
@@ -0,0 +1,79 @@
+---------------------------------------------------------------
+Static typing for ``ndarray`` is a long-term effort that continues 
+with this change.  Arrays are a generic type, parameterized by both 
+shape and data type.  The shape parameter could previously be assigned
```
```suggestion
shape and data type.  The previous shape-type placeholder could previously be assigned
```
Would emphasis here that previously we were truly dealing with a placeholder type slot.

BvB93 reviewed (commented):
Sorry for the delays, got a few mostly documentation-related comments though.

BvB93 commented:
> So I'm coming to believe that any form of npt.Array that uses TypeVarTuple may have to be delayed for another PR. I'm chasing around failed tests, and while things are hunky dory for 3.11 and 3.12, something always goes wrong with 3.10

Does putting the relevant code behind a `if sys.version_info >= (3, 11)` statement make any difference? Usually type checkers are quite good at ignoring code that's outside of such a version bound.

BvB93 commented on numpy/typing/tests/data/fail/shape311.py:
```diff
@@ -0,0 +1,9 @@
+import numpy as np
+import numpy.typing as npt
+
+from typing_extensions import assert_type
```
```suggestion
```
Unused import?

BvB93 commented on numpy/typing/tests/data/fail/shape311.py:
```diff
@@ -0,0 +1,9 @@
+import sys
```
```suggestion
```

ngoldbaum commented on numpy/typing/__init__.py:
```diff
@@ -156,13 +156,14 @@
 )
 
-__all__ = ["ArrayLike", "DTypeLike", "NBitBase", "NDArray"]
+__all__ = ["Array", "ArrayLike", "DTypeLike", "NBitBase", "NDArray"]
```
It looks like introduced a new warning into the numpy doc build:

```
/home/circleci/repo/venv/lib/python3.11/site-packages/numpy/typing/__init__.py:docstring of numpy.typing:220: WARNING: Inline interpreted text or phrase reference start-string without end-string.
```

This is probably caused by the dynamic editing of the docstring below. I'm not sure why adding something now causes this but it was fine before.

Jacob-Stevens-Haas replied:
Thanks!  Yeah it looks like the docs don't build the link `` `npt.Array <numpy.typing.Array>` ``.  Scratching my head but I'm on it.

ngoldbaum reviewed (commented):
I think the other build failure is unrelated to this PR. I re-triggered the build in case it was a transient issue.

There's another issue in the docs build that will need to be fixed, see below. You should be able to run the docs build locally after installing sphinx by doing "spin docs" in the root of the numpy repo.

ngoldbaum commented:
Ah the mac CI failure is due to an issue with spin: https://github.com/scientific-python/spin/issues/193. Unrelated to this PR, please ignore it.

Jacob-Stevens-Haas commented:
Ok, it looks like docs are passing.  Two new CI failures that appear unrelated "E: Unable to locate package g++-13" in dependency steps of Linux SIMD/intel_sde_avx512 and /intel_sde_spr.  Rebased off of main, doesn't clear.

Is this good to go, regardless?

ngoldbaum commented:
Yes those failures are unrelated and if you rebase or merge they'll go away. I haven't been involved with reviewing this and am not familiar with typing so I'll wait for someone else to take a final look.

Jacob-Stevens-Haas commented:
@BvB93 

> Does putting the relevant code behind a if sys.version_info >= (3, 11) statement make any difference? Usually type checkers are quite good at ignoring code that's outside of such a version bound.

Eventually it did.  There was one time I used * instead of `Unpack` in one of the tests, and fixing that made things work.  

> Yes those failures are unrelated and if you rebase or merge they'll go away

I did rebase off of main yesterday.  Rebasing again today just to check, but those two may not go away.  :man_shrugging:

ngoldbaum commented:
There were three different issues that happened to break CI over the past few days and you hit two of them. Should all be good now.

jorenham commented:
The title of this PR doesn't quite cover the load at the moment.

jorenham reviewed (commented):
Because it's currently impossible to have a `typing.TypeVarTuple("ShapeType", covariant=True)`, 
Otherwise, it wouldn't be possible to use e.g. `tuple[Literal[2], Literal[2]]` as a shape type. 
This would be rejected by typecheckers because `Literal[2]` is a subtype of `int`.

Another problem is that the current variadic `Array` type alias is rejected by type checkers (although this could be "solved" by removing the `bound=` of the `ndarray` shape type).

So I'm afraid that the only option is to use a 2-parameter `Array` alias instead, i.e. something like

```python
type Array[ND: tuple[int, ...], SCT: generic] = ndarray[ND, dtype[SCT]]
```

Apart from this having the correct shape type variance and -bound, this also enables `python<3.11` users to use it.

---

FYI, in the [`optype`](https://github.com/jorenham/optype) package there's a reference implementation of such a tuple-based array type alias, specifically, `optype.numpy.Array` ([docs](https://github.com/jorenham/optype?tab=readme-ov-file#arrays)). And I've been using it successfully in several of my (other) projects.

Jacob-Stevens-Haas commented:
> Another problem is that the current variadic Array type alias is rejected by type checkers

In some cases:  
```python
Array[int, int, Any]
```
is accepted
```python
T = TypeVarTuple("T")
Array[*T, Any]
```
is not.  Is that consistent with what you see?  There's was a larger note about this in the documentation, but was reduced to [this](https://github.com/numpy/numpy/blob/403029958d001f37815a5e99a58d7a6f85db35f2/numpy/_typing/_add_docstring.py#L162) after another review.

> 2-parameter `Array` alias 

I definitely test-drove this at some point, and we could do it, but the benefits over just `np.ndarray` typing is a bit limited.  I'd like to hear from @BvB93  or @mtsokol before changing the design again, since they have the power to approve.

jorenham commented:
> > Another problem is that the current variadic Array type alias is rejected by type checkers
> 
> In some cases:
> 
> ```python
> Array[int, int, Any]
> ```
> 
> is accepted
> 
> ```python
> T = TypeVarTuple("T")
> Array[*T, Any]
> ```
> 
> is not. Is that consistent with what you see? 

Which typechecker are you talking about here, and with which flags/settings?

In any case, it would mean that somehow the typechecker implicitly inferred that the (variadic) shape type param of `Array` is bound (when unpacked) to `int`, even though that shouldn't be possible.
So it might be related to a bug, or perhaps some experimental feature.

jorenham commented on numpy/typing/tests/data/pass/shape.py:
```diff
@@ -0,0 +1,23 @@
+    return cast(np.ndarray[tuple[Ax2, Ax1], Any], np.transpose(a))
+
+assert_type(arr.shape, tuple[Time, Series])
+assert_type(transpose(arr).shape, tuple[Series, Time])
```
A common usecase would be to use `typing.Literal` for fixed-size axes. It might be worth testing if that (i.e. the type variance) works as expected here too

Jacob-Stevens-Haas replied:
Done now

Jacob-Stevens-Haas commented:
> Which typechecker are you talking about here, and with which flags/settings?
>
> In any case, it would mean that somehow the typechecker implicitly inferred that the (variadic) shape type param of Array is bound (when unpacked) to int, even though that shouldn't be possible.

@jorenham  I stand corrected.  Originally, I was using both pyright and mypy on one of the many iterations of the test files for this PR.  I had been able to find some use cases that worked and some that didn't, [but was wary of using TypeVarTuple][orig] for the reason you mentioned - inferring the bound shouldn't be possible.  Apparently, something changed in mypy/pyright (or maybe just in doc tests).  Now
```
ndarray[tuple[Unpack[_ShapeTypeTuple]], dtype[_ScalarType_co]]
```
is being highlighted by my syntax highlighter.  Doc tests now fail, but that might be unrelated.

I want to again float the option of _not_ adding an alias.  The original intent of this PR, variance and bounds, are a definite improvement to the state of numpy typing.  Moreover, choosing an alias without `TypeVarTuple` means that when bounds/variance are added to `TypeVarTuple`, we can't introduce an alias that takes advantage of these features.

You've done a lot of work in numpy typing, is there a chance that you could get PR approval authority?  I'm happy to do whatever a numpy approver asks, I just don't want to circle back to an earlier iteration without a clear path to approval.


[orig]: https://github.com/numpy/numpy/pull/26081#issuecomment-2037014185

jorenham commented:
@Jacob-Stevens-Haas It might indeed be a good idea to split this PR into two; one for the (original) covariant shape type parameter, and the other for the `Array` alias. 

Having a covariant shape type is rather important, since otherwise shape typing just isn't going to work. 
Once we have that, numpy (and third party libraries) can start adding type hints for shape types - having a shaped array alias as part of the public typing API isn't necessary for this.

Jacob-Stevens-Haas commented:
Thanks man.  I'm gonna email the list to try to get someone with approval authority to give this attention and make a decision.  If so, it'll only be a small bit of work to move the documentation to the [ndarray section of typing](https://numpy.org/devdocs/reference/typing.html#ndarray), rather than the [API section](https://numpy.org/devdocs/reference/typing.html#api), which currently holds the alias.

Note on the CI: the failure is because doc tests are on 3.10, so any `print(npt.Array)` output is not what is expected. [`typing_extensions.Unpack` isn't really usable in 3.10](https://typing-extensions.readthedocs.io/en/latest/#typing_extensions.Unpack).  Giving the doctests what 3.10 expects would be disingenuous for 3.11+ users reading the docs, who could actually use `npt.Array` with `TypeVarTuple`.

jorenham commented:
> [`typing_extensions.Unpack` isn't really usable in 3.10](https://typing-extensions.readthedocs.io/en/latest/#typing_extensions.Unpack). Giving the doctests what 3.10 expects would be disingenuous for 3.11+ users reading the docs, who could actually use `npt.Array` with `TypeVarTuple`. I can add a conditional `pytest.skip`, if that's really desired.

Yet another good reason (besides those I mentioned in https://github.com/numpy/numpy/pull/26081#pullrequestreview-2170708862) to replace the `TypeVarTuple(_)` with a `TypeVar(_, bound=tuple[int, ...], covariant=True)` 😉

ngoldbaum commented:
Would you mind coming to the next [PR triage meeting this Wednesday](https://scientific-python.org/calendars/numpy.ics) to talk about this PR? You can also add it to the agenda if you can't make it.

Jacob-Stevens-Haas commented:
> Would you mind coming to the next [PR triage meeting this Wednesday](https://scientific-python.org/calendars/numpy.ics) to talk about this PR? You can also add it to the agenda if you can't make it.

Thanks, I'll add it to the agenda and my calendar

rgommers commented:
> Thanks man. I'm gonna email the list to try to get someone with approval authority to give this attention and make a decision.

Given that @BvB93 currently has low availability, and @jorenham is doing a ton of great work on static typing (hard to keep up with!) and is actively reviewing here, I'd be happy to rely on @jorenham as the main reviewer here. If the two of you are both happy with the state of this PR (or two PRs, if you split off the alias), I'd be happy to hit the merge button if needed.

I'll reply on the mailing list as well.

And thank you for sticking with this PR! I'm very aware that static typing improvements & handling shapes better is impactful and we have a gap between user demand and what we've able to do over the last couple of years. Great to see this much activity recently.

Jacob-Stevens-Haas commented:
Great, thanks @rgommers!  @jorenham I'll remove the alias, review the doc changes, fix the CI, and re-request review.

Jacob-Stevens-Haas commented:
Alright @jorenham, ready for review!  Updated built documentation on [ndarray typing](https://output.circle-artifacts.com/output/job/8419d824-7ccb-46c1-9d40-2b16c27e5971/artifacts/0/doc/build/html/reference/typing.html#ndarray) and [release notes](https://output.circle-artifacts.com/output/job/8419d824-7ccb-46c1-9d40-2b16c27e5971/artifacts/0/doc/build/html/release/2.1.0-notes.html#ndarray-shapes-are-now-covariant-and-bound)

jorenham commented on numpy/__init__.pyi:
```diff
@@ -1651,7 +1651,7 @@ class ndarray(_ArrayOrScalarCommon, Generic[_ShapeType, _DType_co]):
-    def shape(self) -> _Shape: ...
+    def shape(self) -> _ShapeType: ...
     @shape.setter
     def shape(self, value: _ShapeLike) -> None: ...
```
This setter is a bit of a problem...

Because from typing perspective, we can't just change the shape type from e.g. `tuple[int]` to `tuple[int, int]`, even though at runtime we can

```
x = np.arange(42)
x.shape = (7, 6)
```

But if we were to only accept `value: _ShapeType` here, then `_ShapeType` cannot be covariant anymore, which kinda defeats the purpose of this entire PR.
And I'm not sure that ignoring the resulting mypy errors about invalid typevar use, is wise.

So I think that the most reasonable option is to remove this `shape` setter (only in the stubs).
Perhaps with a comment such as `The shape type is covariant and should not be written to; use .reshape() instead`.

@seberg , what do you think?

Jacob-Stevens-Haas replied:
I'm +1 on removing the stub for setting shape

jorenham commented on numpy/_core/_add_newdocs.py:
```diff
@@ -3092,7 +3092,6 @@
     :pep:`585` : Type hinting generics in standard collections.
     numpy.typing.NDArray : An ndarray alias :term:`generic <generic type>`
                         w.r.t. its `dtype.type <numpy.dtype.type>`.
-
```
This whitespace fix seems out-of-scope for this PR.

Jacob-Stevens-Haas replied:
Gotcha, that was just a leftover from adding an entry for `npt.Array` and then removing it, rather than reverting it.

jorenham commented on numpy/_typing/_add_docstring.py:
```diff
@@ -3,7 +3,7 @@
 import textwrap
 
-from ._array_like import NDArray
+from numpy._typing._array_like import NDArray
```
Could you revert this?

Jacob-Stevens-Haas replied:
Yeah my bad there

jorenham commented on numpy/_typing/_add_docstring.py:
```diff
@@ -120,7 +120,7 @@ def _parse_docstrings() -> str:
 add_newdoc('NDArray', repr(NDArray),
     """
-    A `np.ndarray[Any, np.dtype[+ScalarType]] <numpy.ndarray>` type alias 
+    A `np.ndarray[Any, np.dtype[+ScalarType]] <numpy.ndarray>` type alias
```
Another out-of-scope whitespace fix. 

(If you're using vscode, then I believe you can configure it to only strip trailing whitespaces for the lines that you actually edited, or something like that)

jorenham commented on numpy/ma/core.pyi:
```diff
@@ -17,7 +17,7 @@ from numpy import (
 # TODO: Set the `bound` to something more suitable once we
 # have proper shape support
-_ShapeType = TypeVar("_ShapeType", bound=Any)
+_ShapeType = TypeVar("_ShapeType", covariant=True, bound=tuple[int, ...])
```
Not the most interesting issue, but the parameter order of `covariant` and `bound` is exactly the opposite as the line below, which to me just feels kinda "wrong" 😛.

```suggestion
_ShapeType = TypeVar("_ShapeType", bound=tuple[int, ...], covariant=True)
```

jorenham replied:
And here it should also be better to add a `_co` suffix to the name.

jorenham commented on numpy/ma/core.pyi:
```diff
@@ -17,7 +17,7 @@ from numpy import (
 
 # TODO: Set the `bound` to something more suitable once we
 # have proper shape support
```
```suggestion
```

jorenham commented on numpy/ma/mrecords.pyi:
```diff
@@ -7,7 +7,7 @@ __all__: list[str]
 
 # TODO: Set the `bound` to something more suitable once we
 # have proper shape support
```
```suggestion
```

jorenham commented on numpy/ma/mrecords.pyi:
```diff
@@ -7,7 +7,7 @@ __all__: list[str]
 # TODO: Set the `bound` to something more suitable once we
 # have proper shape support
-_ShapeType = TypeVar("_ShapeType", bound=Any)
+_ShapeType = TypeVar("_ShapeType", covariant=True, bound=tuple[int, ...])
```
```suggestion
_ShapeType = TypeVar("_ShapeType", bound=tuple[int, ...], covariant=True)
```

jorenham replied:
A `_co` suffix would help here as well

jorenham commented on numpy/typing/__init__.py:
```diff
@@ -64,6 +64,31 @@
 ndarray
 ~~~~~~~
 
+The ``ndarray`` type is generic with respect to both its shape and data type.
+The shape parameter is a tuple of ints or subtypes of int.
+
+The data type field is a subclass of ``np.dtype``, usually via parametrization
+with a scalar type: e.g.
+``Float1D = np.ndarray[tuple[int], np.dtype[np.float64]]``.
+
+Array creation cannot yet detect the shape parameter from arguments; it must be
+declared, e.g.:
+
+.. code-block:: python
+
+    > from typing import Literal, NewType
+    > Width = NewType("Width", int)
+    > Height = NewType("Height", int)
+    >
+    > GrayImg = np.ndarray[tuple[Width, Height], np.dtype[np.uint8]]
+    > RGBImg = np.ndarray[tuple[Width, Height, Literal[3]], np.dtype[np.uint8]]
+    >
+    > def colorize(img: GrayImg) -> RGBImg: ...
+    > 
+    > x: GrayImg = np.array([[1, 2, 3], [4, 5, 6]])
+    > assert_type(colorize(x), RGBImg)  # type check passes
+
+
```
This looks very similar to the text change notes, so my comments there also apply here.

Jacob-Stevens-Haas replied:
Following those comments from above, if I remove the detail from the change notes, are you ok with a more detailed explanation here?

jorenham replied:
I don't think this is needed at all actually, as this is directed to the few of us that work on numpy typing.
If we're unsure of the specifics of the shape-type-parameter, then we can just look at the `TypeVar` itself (at least, that's what I'd do; but that's probably because my attention span is in units of attoseconds).

Jacob-Stevens-Haas replied:
That all makes sense, but until a week ago, I wasn't really planning on working more deeply with numpy typing.  I was exactly the user you described above: one who sees this change as a feature that allows me to shape-type my own functions, even if I have to `cast` the output.

I'm ok removing the docs for now, though.  I agree it could be confusing for users who, as soon as they do something like `arr + 1`, lose the shape type information.

jorenham commented on numpy/typing/tests/data/pass/shape.py:
```diff
@@ -0,0 +1,28 @@
+from typing import Any, cast, Literal, NewType, TypeVar
```
These tests look more like they belong in `numpy/typing/test/data/reveal`.

And because those are `.pyi` tests, the runtime code doesn't matter anymore

jorenham commented on numpy/typing/tests/data/pass/shape.py:
```diff
@@ -0,0 +1,28 @@
+import numpy as np
+from typing_extensions import assert_type
+
+Time = NewType("Time", int)
+Series = NewType("Series", int)
```
My earlier comment on `NewType` also applies here for the most part. 
Just using `int` directly should be enough.

jorenham commented on numpy/typing/tests/data/pass/shape.py:
```diff
@@ -0,0 +1,28 @@
+# Test literal
+arr2: np.ndarray[tuple[Literal[1], Literal[2]], Any] = np.arange(2).reshape((1, 2))
+assert_type(arr2.shape, tuple[Literal[1], Literal[2]])
+assert_type(transpose(arr2).shape, tuple[Literal[2], Literal[1]])
```
This also tests (co)variance of the tuple values like the test on line 20. So one of them is redundant.

jorenham replied:
There are tests for the covariance of the tuple *values*, but not yet for the covariance of the tuple *itself*. So this doesn't actually test the covariance of the shape type. 
So e.g. passing some `typing.NamedTuple` instance as a shape.

BTW, you might wanna check whether this'll match with the runtime behavior (I don't mean as a test, just in a python shell or something)

Jacob-Stevens-Haas replied:
Oooh, interesting point, and you're right.  I was less familiar with using `Literal`, but apparently it is treated as a subtype in a way "identical to how we handle NewTypes."

I think both do handle covariance of `tuple` itself, because tuple is covariant with respect to _its_ arguments - `tuple[Time, Literal[2]]` is a subtype of `tuple`.  But you're right it makes sense to add a test using a real tuple subclass.

Jacob-Stevens-Haas replied:
If you meant that `reshape` works with a namedtuple of ints, then yeah, just checked it :heavy_check_mark:

jorenham replied:
> On the last part, I'm not sure what you mean. What runtime behavior is uncertain?

```pycon
>>> import numpy as np
>>> from typing import NamedTuple
>>> class Shape2D(NamedTuple):
...     x: int
...     y: int
... 
>>> x = np.empty(Shape2D(6, 7))
>>> type(x.shape)
<class 'tuple'>
```

Jacob-Stevens-Haas replied:
Ah! It would be cool.  Alas

jorenham replied:
I've opened an issue about it: https://github.com/numpy/numpy/issues/27086

jorenham commented on doc/release/upcoming_changes/26081.improvement.rst:
```diff
@@ -0,0 +1,41 @@
+``ndarray`` Shapes are now covariant and bound
```
```suggestion
``ndarray`` shape-type parameter is now covariant and bound to ``tuple[int, ...]``
```

Jacob-Stevens-Haas commented on doc/release/upcoming_changes/26081.improvement.rst:
```diff
@@ -0,0 +1,41 @@
+   from typing import Any, NewType, TypeVar
+   import numpy as np
+
+   Time = NewType("Time", int)
+   Series = NewType("Series", int)
+   DType = TypeVar("DType", bound=np.dtype)
+   Ax1 = TypeVar("Ax1", bound=int)
+   Ax2 = TypeVar("Ax2", bound=int)
+
+   arr: np.ndarray[tuple[Time, Series], Any] = np.arange(4.0).reshape((2, 2))
```
If I understand correctly, type declarations are valid so long as they narrow the inferred type.  Since no array creation routines infer the `_ShapeType` from it's arguments,
```
arr: np.ndarray[tuple[Time, Series], Any] = np.arange(4.0).reshape((2, 2))
```
is valid, as is 
```
arr: np.ndarray[tuple[Literal[2], Literal[2]], Any] = np.arange(4.0).reshape((2, 2))
```
because they both narrow the inferred type `ndarray[Any, dtype[floating[Any]]` and obey the type bounds for shape and dtype.  I agree that they can't be combined, as `Literal[Time(2)]` isn't valid.  It's an either or: you can use literal length axes, or you can use `NewType` to say that the axes mean something specific (e.g `Time`, `Series`).  Functions that require one will reject an array declared as the other unless it is later `cast`.

For me, the original point of the PR/covariance was to use `NewType`, as `int` is rarely going to have any true subclasses.  [The example in PEP 646](https://peps.python.org/pep-0646/#motivation) is pretty close to my use case: a library that handles images and videos, and the need for type checking to handle the difference between a gray movie (with axes height, width, frame) from an RGB image (with axes height, width, channel).  E.g. a function `lower_resolution()` should persist its input type and could apply to both, but a function `convert_gray()` that averages the last axis should only accept an RGB image (or RGB movie).

But I agree with your next comment, that an in-depth example is better in the `numpy.typing` docs, rather than release notes.  How about I shrink this example to just show declarations of `int`, `Literal` and `NewType`?

jorenham replied:
> For me, the original point of the PR/covariance was to use NewType, as int is rarely going to have any true subclasses.

`Literal[N: int]` is a subtype of `int`, and it has many use-caaes, e.g.

```python
type ColumnVector[N: int, T: number[Any]] = ndarray[tuple[Literal[1], N], dtype[T]]
type RotationMatrix[ND: Literal[2, 3], T: floating[Any]] = ndarray[tuple[ND, ND], dtype[T]]
type ImageRGB[W: int, H: int] = ndarray[tuple[H, W, Literal[3]], dtype[uint8]]
```

But that's not all that this PR solves.
NumPy has many examples of functions to accept arrays with either 1 or 2 dimensions,
e.g.:

```python
def f(x: ndarray[tuple[int] | tuple[int, int], dtype[Any]): ...
```

But currently you can't pass `np.ndarray[tuple[int], np.dtype[Any]]`.
If you try do do so, then Pyright reports:

```
Argument of type "ndarray[tuple[int], dtype[Any]]" cannot be assigned to parameter "x" of type "ndarray[tuple[int] | tuple[int, int], dtype[Any]]" in function "g"
  "ndarray[tuple[int], dtype[Any]]" is incompatible with "ndarray[tuple[int] | tuple[int, int], dtype[Any]]"
    Type parameter "_ShapeType@ndarray" is invariant, but "tuple[int]" is not the same as "tuple[int] | tuple[int, int]"
```

This is actually the reason why shape-typing in NumPy is practically impossible at the moment.

---

To my knowledge, there aren't any other issues about these problems specifically.
The way I see it, is that this change allows us to start adding shape-type annotations within NumPy, so this is not a new feature or a bugfix.

Also note that all numpy array constructors currently return `Any` for the shape, and shape-types are almost never used in general.
So unless users currently explicitly `cast` an array to have shape-types, or use a custom shape-typed array constructor or something, this change just isn't *directly* relevant for them.

So I don't think an example is needed.
This is not a feature and not a bugfix; this opens the road to the long-awaited shape-typing of arrays 🎉.

So I don't think an example is really necessary here.
Plus, showing an example of shape-typing in the release notes, even though there's almost no support for it in numpy, is bound to confuse a lot of people.

Jacob-Stevens-Haas commented on numpy/__init__.pyi:
```diff
@@ -1502,8 +1502,8 @@ _FlexDType = TypeVar("_FlexDType", bound=dtype[flexible])
-_ShapeType = TypeVar("_ShapeType", bound=Any)
-_ShapeType2 = TypeVar("_ShapeType2", bound=Any)
+_ShapeType = TypeVar("_ShapeType", covariant=True, bound=tuple[int, ...])
+_ShapeType2 = TypeVar("_ShapeType2", covariant=True, bound=tuple[int, ...])
```
I'm confused by this one - I'm happy to make the change (`__array_wrap__` is sort of deprecated), but if it's used as a shape type, why wouldn't it be covariant?  And if, as you say, it's not bound to any type, why keep the bound?

jorenham replied:
When a type parameter is co/contra-variant, it is always with respect to a generic *type*.
For generic functions, there is no such thing as a covariant type parameter; there's nothing to vary.

Taking function parameters as analogue, it'd be a bit like taking the keyword-only `out` parameter of a specific `ufunc`, and re-using it as a method parameter of some unrelated type. 

Both function parameters and type parameters can be seen as (contextual) placeholders.
But where function parameters accept *values* or *instances*, type parameters accept only, and suprisingly, *types* (plus they live in the typing-dimension, as opposed to the runtime one, and they're interconnected with wormholes, but I digress).

BTW, that's one of the reasons why I often use Python 3.12 syntax when giving typing examples, as it doesn't have these problems, and the variance of type params is automagically inferred (even though it can't always read your mind it turns out, but let's not go there now).

Jacob-Stevens-Haas replied:
I don't understand all of that, but it definitely helps me get partway through the wormhole.  I'm happy to make the change.  And yeah, python 3.12 syntax does make variable scope much clearer :)

Jacob-Stevens-Haas commented:
Thanks Joren - good catches all around.  My only thing is that `NewType` and `Literal` both seem important - but I agree that the details belong in the numpy.typing docs, rather than the release notes.

jorenham commented on numpy/typing/tests/data/reveal/shape.pyi:
```diff
@@ -0,0 +1,28 @@
+# Test bound
+arr: np.ndarray[tuple[Time, Series], Any] = np.arange(4.0).reshape((2, 2))
+
+Ax1 = TypeVar("Ax1", bound=int)
+Ax2 = TypeVar("Ax2", bound=int)
+def transpose(
+    a: np.ndarray[tuple[Ax1, Ax2], Any]
+) -> np.ndarray[tuple[Ax2, Ax1], Any]:
+    return cast(np.ndarray[tuple[Ax2, Ax1], Any], np.transpose(a))
```
See https://github.com/python/typing/issues/548 for some concrete examples

purepani replied:
Oh wait yeah.

Too many layers of abstraction :)
Sorry about the noise

jorenham replied:
No worries; it's always good to see interest in python typing :)

Jacob-Stevens-Haas replied:
Hey purepani, yes, the test passes.  I'm not sure what you mean by "higher kindled type", but this part of the test would've passed even pre-PR.  Which maybe suggests it's not the best test.

Jacob-Stevens-Haas commented:
Alright, I think I've got all of the comments now. If it passes tests, it should be good to re-review.  Do you prefer if reviewer or reviewee resolves old review comments?  I tend to leave them open with a :+1: once I think we're on the right page in case you want to check them in re-review, but I know people differ on this.

jorenham commented on numpy/typing/tests/data/reveal/shape.pyi:
```diff
@@ -0,0 +1,20 @@
+# Test variance of _ShapeType_co
+def accepts_2d(a: np.ndarray[tuple[int, int], Any]) -> None: ...
+
+accepts_2d(arr)
```
Since this doesn't "reveal" anything, it makes more sense to place this one in a `pass/shape.py`.

Alternatively, you could transform `accepts_2d` to e.g. return the shape of the input array, so that you could do `assert_type(accepts_2d(arr), XYGrid)`.

jorenham commented on numpy/__init__.pyi:
```diff
@@ -1502,8 +1502,8 @@ _FlexDType = TypeVar("_FlexDType", bound=dtype[flexible])
 
 # TODO: Set the `bound` to something more suitable once we
 # have proper shape support
```
```suggestion
```

jorenham commented:
> Do you prefer if reviewer or reviewee resolves old review comments?

I think it'll make reviewing easier if I mark them as resolved myself

jorenham reviewed (changes requested):
There are only a couple of minor suggestions left now.

Once those are out of the way, I think that we can go ahead and merge this :)

---

**edit:**

I noticed that the commit history has gotten quite messy, and should probably be cleaned up a bit, as described in: 
https://numpy.org/doc/stable/dev/development_workflow.html#rewriting-commit-history

Jacob-Stevens-Haas commented on numpy/typing/tests/data/pass/shape.py:
```diff
@@ -0,0 +1,18 @@
+    x_axis: int
+    y_axis: int
+
+arr: np.ndarray[XYGrid, Any] = np.array([[1, 2], [2, 3]])
```
I'm happy to do that, but why can't it be `XYGrid`?  the `array` constructor sets the shape type parameter to `Any`

jorenham replied:
Becaase `np.array` returns a shape of type `tuple[int, ...]` by default, which isn't a subtype of `XYGrid`.

Jacob-Stevens-Haas commented:
> I noticed that the commit history has gotten quite messy, and should probably be cleaned up a bit

Would you be ok with me just squashing all commits together?  It's not like there's several pieces put together here - just one change, copied across a few subclasses, the tests for that change, and the release note for that change.

jorenham commented:
> > I noticed that the commit history has gotten quite messy, and should probably be cleaned up a bit
> 
> Would you be ok with me just squashing all commits together? It's not like there's several pieces put together here - just one change, copied across a few subclasses, the tests for that change, and the release note for that change.

Yea that's fine by me

Jacob-Stevens-Haas committed:
TYP: Make array shape type variable covariant and bound

Fixes #25729

This change allows future changes to the static typing of numpy that modify
or only work with certain numbers of dimensions.  It also applies the change
to subclasses of ndarray and adds tests.  It allows users to statically
type their array shapes with subtypes of tuple (e.g. NamedTuple) and tuples
of int subtypes (e.g. Literal or NewType).

For a discussion of the merits of TypeVarTuple vs a tuple-bound TypeVar, see
the linked PR

jorenham commented on .gitignore:
```diff
@@ -1,5 +1,6 @@
 # Editor temporary/working/backup files #
 #########################################
+env/
```
🤔

Jacob-Stevens-Haas replied:
Oops yeah that wasn't supposed to be committed

jorenham commented on doc/release/upcoming_changes/26081.improvement.rst:
```diff
@@ -0,0 +1,10 @@
+the shape and the data type.  Previously, the shape type parameter could be
+any value.  This change restricts it to a tuple of ints, as one would expect
+from using ``ndarray.shape``.  Further, the shape-type parameter has been
+changed from invariant to covariant.  This change also applies to subpytes of
+``ndarray``, e.g. ``np.ma.MaskedArray``.  See the `typing docs <https://typing.readthedocs.io/en/latest/reference/generics.html#variance-of-generic-types>`_
```
```suggestion
changed from invariant to covariant.  This change also applies to the subtypes of
``ndarray``, e.g. ``numpy.ma.MaskedArray``.  See the `typing docs <https://typing.readthedocs.io/en/latest/reference/generics.html#variance-of-generic-types>`_
```

jorenham commented on numpy/__init__.pyi:
```diff
@@ -3927,7 +3925,7 @@ class poly1d:
     ) -> poly1d: ...
 
-class matrix(ndarray[_ShapeType, _DType_co]):
+class matrix(ndarray[_ShapeType_co, _DType_co]):
```
It might be kinda neat to narrow this one to `bound=tuple[int, int]`.
But feel free to ignore/resolve this if you don't want to; I wouldn't mind merging it like this :)

jorenham reviewed (changes requested):
I left two tiny suggestions, and a bonus one.
Once these two are fixed, we can merge this AFAIK

Jacob-Stevens-Haas committed:
TYP: Force matrix shape type to be 2D

jorenham commented on numpy/__init__.pyi:
```diff
@@ -3927,7 +3925,9 @@ class poly1d:
 
-class matrix(ndarray[_ShapeType, _DType_co]):
+
+_ShapeType_2D_co = TypeVar("_ShapeType_2D_co", covariant=True, bound=tuple[int, int])
```
I think it makes sense to move this towards the other `_ShapeType` type vars.

But you can leave it here as well, although you should then add a newline after it, since classes in `.pyi` files require at least 1 newline before/after it (if I remember correctly).

jorenham replied:
The current name might also be a bit confusing with `_ShapeType2`.
Renaming it to something like `_MatrixShapeType_co` or `_Shape2DType_co` could help with this.

jorenham commented on doc/release/upcoming_changes/26081.improvement.rst:
```diff
@@ -0,0 +1,11 @@
+the shape and the data type.  Previously, the shape type parameter could be
+any value.  This change restricts it to a tuple of ints, as one would expect
+from using ``ndarray.shape``.  Further, the shape-type parameter has been
+changed from invariant to covariant.  This change also applies to the subpytes
```
you forgot about the "subpytes" ~pyto~ typo :)

```suggestion
changed from invariant to covariant.  This change also applies to the subtypes
```

jorenham commented:
Thanks a lot @Jacob-Stevens-Haas! Everyone can now go ahead and start typing the shapes of arrays 🎉.

I'm looking forward to your PR for the shaped `Array` type alias -- assuming that this one didn't scare you off, that is 😜.

Jacob-Stevens-Haas commented:
Thanks again for your help and enthusiasm @jorenham!  I'm curious whether you think it's worth working on the PEP for `TypeVarTuple` variance/bounds before the alias.  The challenge being that we can't come back to the ideal alias in a backwards-compatible manner if we choose a non-ideal alias now.  I started to look at the implementation of `TypeVarTuple` in CPython, and it could be an achievable goal (there's a `BoundsVarianceMixin` in the `typing` module that `TypeVar` uses, maybe that's compatible?)

jorenham commented:
> I'm curious whether you think it's worth working on the PEP for `TypeVarTuple` variance/bounds ...

Please do! Perhaps you could also allow `ParamSpec` to be co/contra-variant while you're at it 🤔; it's a very similar case.

> ... before the alias. 

Well, besides the covariance and bounds issues, there are other reasons why I think that a plain tuple `TypeVar` is a better fit in this case. 
A big one is that converting from a `ShapeT: tuple[int, ...]` to `*ShapesT: int` is a one-way street. 

So this is not a problem:

```python
type ArrayTup[ShapeT: tuple[int, ...], SCT: generic] = ndarray[ShapeT, dtype[SCT]]
type ArrayVar[*Shape: int, SCT: generic] = ArrayTup[tuple[*Shape], SCT]
```

But the other way around is impossible:

```python
type ArrayVar[*Shape: int, SCT: generic] = ndarray[tuple[*Shape], dtype[SCT]]
type ArrayTup[ShapeT: tuple[int, ...], SCT: generic] = ArrayVar[???, SCT]
```

---

> I started to look at the implementation of `TypeVarTuple` in CPython, and it could be an achievable goal (there's a `BoundsVarianceMixin` in the `typing` module that `TypeVar` uses, maybe that's compatible with `TypeVarTuple`?)

Don't underestimate the amount of bureaucracy and discussion that's involved in attempting to change things like this:

I once made an attempt at a similar PEP that allowed specifying co- and contra- variant type params as `+T` and `-T` (like in e.g. Scala), which got rejected (because of PEP 695 and its magic `infer_variance`, which turns out it horrible for debugging, and not always right). But even so, it already was a lot of work, and that was *before* the official typing specification existed.

I'm not trying to demotivate you or anything, but just know what you're getting yourself into. 
As it currently stands, `TypeVarTuple` and `ParamSpec` are useless in most cases for which they were meant. 
So it would be of great help if you could manage to revive them.
And even if you won't be able to, it's already worthwhile if you can get people to start talking about this -- it currently isn't considered to be much of a problem at all, as far as I'm aware.

rgommers commented:
Great to see this land (and just in time for 2.1.0 too), thanks a lot @Jacob-Stevens-Haas, @jorenham & all reviewers!
